### PR TITLE
RAINCATCH-539 - angular mediator

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation 

Pull out angular source code to separate module. Use mediator to call client side code.

Review @witmicko @nialldonnellyfh @JameelB

### CODE REVIEW

**Comment by TommyJ1994** [comment]
@wtrocki Can you bump `express` in package.json to `4.15.2` in this PR? I'm seeing failing builds, and it might be because the master branch is not currently in a working state (this PR might be addressing the issue)?

You can fix the failing build for you PR by adding this in `.travis.yml`
```
before_install:
    - npm install -g grunt-cli
```

File: lib/mediator-service/index.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ * @param {object}   config
+ * @constructor
+ */
+function FileMediatorService(mediator, config) {
+  this.mediator = mediator;
+  this.config = config || {};
+  
+  this.usersTopics = new MediatorTopicUtility(mediator).prefix(CONSTANTS.TOPIC_PREFIX).entity(CONSTANTS.USERS_ENTITY_NAME);
+  this.filesTopics = new MediatorTopicUtility(mediator).prefix(CONSTANTS.TOPIC_PREFIX).entity(CONSTANTS.FILES_ENTITY_NAME);
+  this.fileUITopics = new MediatorTopicUtility(mediator).prefix(CONSTANTS.TOPIC_PREFIX).entity(CONSTANTS.FILE);
```
Comment (JameelB) [inline lib/mediator-service/index.js]: @wtrocki should this not be using the `FILE_TOPIC_UI_PREFIX`? https://github.com/wtrocki/raincatcher-file-angular/blob/68ec0c752cd0edcdacdaeee97c853c5b8d1da4c2/lib/constants.js#L7

File: lib/file-client/index-spec.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+describe("Test index", function() {
+  beforeEach(function() {
+  });
+  
+  afterEach(function() {
+  });
+  
+  it('test', function() {
+  });
+});
```
Comment (JameelB) [inline lib/file-client/index-spec.js]: Empty tests.

File: lib/file-client/index-spec.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+var Client = require('./index');
+var should = require('should');
```
Comment (JameelB) [inline lib/file-client/index-spec.js]: This dependency was never added to package.json

File: lib/file-client/index-spec.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+var fileSubscribers = new MediatorTopicUtility(mediator);
+fileSubscribers.prefix(CONSTANTS.TOPIC_PREFIX).entity(CONSTANTS.FILES_ENTITY_NAME);
+
+describe("Client tests", function() {
+  beforeEach(function() {
+    this.subscribers = {};
+  });
+  
+  afterEach(function() {
```
Comment (JameelB) [inline lib/file-client/index-spec.js]: empty afterEach?

File: README.md [hunk 5]
```diff
@@ -1,24 +1,62 @@
...
+var config = {
+  uploadEnabled: false,
+  listColumnViewId: "column2",
+  mainColumnViewId: "content@app",
+  detailStateMount: "app.file.detail"
+}
+
 angular.module('app', [
-, require('fh-wfm-file')
+, require('fh-wfm-file')(config)
```
Comment (JameelB) [inline README.md]: this should be `require('fh-wfm-file-angular')(config)`

**Review by JameelB — DISMISSED** [review]

File: lib/file-client/index.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+var CONSTANTS = require('../constants');
+var MediatorTopicUtility = require('fh-wfm-mediator/lib/topics');
+var mediator, manager, fileSubscribers;
+
+/**
+ * Getting Promises for done and error topics.
+ *
+ * @param doneTopicPromise  - A promise for the done topic.
+ * @param errorTopicPromise - A promise for the error topic.
+ * @returns {*}
```
Comment (witmicko) [inline lib/file-client/index.js]: missing param type, http://usejsdoc.org/tags-param.html name and type

File: lib/file-client/index.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    };
+  });
+}
+
+/**
+ *
+ * Initialising the file-client with a mediator.
+ *
+ * @param _mediator
+ * @returns {ManagerWrapper|*}
```
Comment (witmicko) [inline lib/file-client/index.js]: missing param type http://usejsdoc.org/tags-param.html

File: lib/file-detail/file-detail-controller.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+var _ = require("lodash");
+var $q = require("q");
+
+var module = angular.module(CONSTANTS.FILE_DIRECTIVE_MODULE);
+
+// View controller for current implementation
+module.controller('FileDetailController', function($scope, $state, $stateParams, userClient, mediator, fileMediatorService, $timeout){
+  var self = this;
+  self.closeFile = function(event){
+    mediator.publish('wfm:file:detail:close');
```
Comment (witmicko) [inline lib/file-detail/file-detail-controller.js]: why not use constants for topic?

File: lib/file-detail/file-detail-controller.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    event.preventDefault();
+    event.stopPropagation();
+  };
+  
+  mediator.subscribeForScope('wfm:file:detail:close', $scope, function() {
+    $state.go('app.file');
+  });
+
```
Comment (witmicko) [inline lib/file-detail/file-detail-controller.js]: topic constants

File: lib/mediator-service/index.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+ * Getting Promises for done and error topics.
+ * This will resolve or reject the returned promise depending on the topic published.
+ *
+ * TODO: This may be of more use in fh-wfm-mediator...
+ *
+ * @param doneTopicPromise  - A promise for the done topic.
+ * @param errorTopicPromise - A promise for the error topic.
+ * @returns {Promise}
```
Comment (witmicko) [inline lib/mediator-service/index.js]: missing param type

File: lib/service.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+angular.module(CONSTANTS.FILE_MODULE_ID, [])
+  .factory('fileService', function($q, $timeout, mediator){
+    var fileService = {};
+    var fileClient = fileClient(mediator);
+    _.forOwn(fileClient, function(value, key){
+      if(typeof value === 'function'){
+        fileService[key] = function(){
+          return $q.when(fileClient[key].apply(fileClient, arguments));
+        };
+      }else{
```
Comment (witmicko) [inline lib/service.js]: missing spaces around {} ?

**Review by witmicko — COMMENTED** [review]
looks good, few minor comments

**Comment by ghost** [comment]
LGTM

**Pull request merged**